File: android/app/src/main/kotlin/com/meshlink/meshlink/model/MeshEvent.kt
package com.meshlink.meshlink.model

import com.meshlink.meshlink.constants.MeshConstants

sealed class MeshEvent {
    abstract val eventType: String
    abstract fun toMap(): Map<String, Any?>

    data class PeerFound(
        val peerId: String,
        val peerName: String,
        val rssi: Int = -60,
        val connectionType: String = "BLE"
    ) : MeshEvent() {
        override val eventType: String = MeshConstants.EVENT_PEER_FOUND
        override fun toMap(): Map<String, Any?> = mapOf(
            "eventType" to eventType,
            "peerId" to peerId,
            "peerName" to peerName,
            "rssi" to rssi,
            "connectionType" to connectionType
        )
    }

    data class PeerLost(val peerId: String) : MeshEvent() {
        override val eventType: String = MeshConstants.EVENT_PEER_LOST
        override fun toMap(): Map<String, Any?> = mapOf(
            "eventType" to eventType,
            "peerId" to peerId
        )
    }

    data class PeerConnected(
        val peerId: String,
        val groupOwnerIp: String? = null,
        val isGroupOwner: Boolean = false
    ) : MeshEvent() {
        override val eventType: String = MeshConstants.EVENT_PEER_CONNECTED
        override fun toMap(): Map<String, Any?> = mapOf(
            "eventType" to eventType,
            "peerId" to peerId,
            "groupOwnerIp" to groupOwnerIp,
            "isGroupOwner" to isGroupOwner
        )
    }

    data class PeerDisconnected(val peerId: String) : MeshEvent() {
        override val eventType: String = MeshConstants.EVENT_PEER_DISCONNECTED
        override fun toMap(): Map<String, Any?> = mapOf(
            "eventType" to eventType,
            "peerId" to peerId
        )
    }

    data class PacketReceived(val packetJson: String) : MeshEvent() {
        override val eventType: String = MeshConstants.EVENT_PACKET_RECEIVED
        override fun toMap(): Map<String, Any?> = mapOf(
            "eventType" to eventType,
            "packetJson" to packetJson
        )
    }

    /**
     * Emitted for global connection state changes (existing fields) AND
     * per-peer Wi-Fi Direct state changes (peerId + connectionState).
     * Additive — existing Flutter consumers reading only isConnected/activePeersCount
     * are unaffected by the new optional fields.
     */
    data class ConnectionStateChanged(
        val isConnected: Boolean,
        val activePeersCount: Int,
        val peerId: String? = null,
        val connectionState: String? = null  // "connecting" | "connected" | "failed" | "disconnected"
    ) : MeshEvent() {
        override val eventType: String = MeshConstants.EVENT_CONNECTION_STATE_CHANGED
        override fun toMap(): Map<String, Any?> = mapOf(
            "eventType" to eventType,
            "isConnected" to isConnected,
            "activePeersCount" to activePeersCount,
            "peerId" to peerId,
            "connectionState" to connectionState
        )
    }

    data class MeshError(val errorCode: String, val message: String) : MeshEvent() {
        override val eventType: String = MeshConstants.EVENT_ERROR
        override fun toMap(): Map<String, Any?> = mapOf(
            "eventType" to eventType,
            "errorCode" to errorCode,
            "message" to message
        )
    }
}

File: lib/domain/entities/mesh_event.dart
import '../../data/models/packet_model.dart';

sealed class MeshEvent {
  const MeshEvent();

  factory MeshEvent.fromMap(Map<String, dynamic> map) {
    final eventType = map['eventType'] as String? ?? '';
    switch (eventType) {
      case 'peerFound':
        return PeerFoundEvent(
          peerId: map['peerId'] as String? ?? '',
          peerName: map['peerName'] as String? ?? 'Unknown Peer',
          rssi: (map['rssi'] as num?)?.toInt() ?? -70,
          connectionType: map['connectionType'] as String? ?? 'BLE',
        );
      case 'peerLost':
        return PeerLostEvent(peerId: map['peerId'] as String? ?? '');
      case 'peerConnected':
        return PeerConnectedEvent(
          peerId: map['peerId'] as String? ?? '',
          groupOwnerIp: map['groupOwnerIp'] as String?,
          isGroupOwner: map['isGroupOwner'] as bool? ?? false,
        );
      case 'peerDisconnected':
        return PeerDisconnectedEvent(peerId: map['peerId'] as String? ?? '');
      case 'packetReceived':
        final jsonStr = map['packetJson'] as String? ?? '{}';
        return PacketReceivedEvent(packet: PacketModel.fromJson(jsonStr));
      case 'connectionStateChanged':
        return ConnectionStateChangedEvent(
          isConnected: map['isConnected'] as bool? ?? false,
          activePeersCount: (map['activePeersCount'] as num?)?.toInt() ?? 0,
          peerId: map['peerId'] as String?,
          connectionState: map['connectionState'] as String?,
        );
      case 'error':
        return MeshErrorEvent(
          errorCode: map['errorCode'] as String? ?? 'UNKNOWN',
          message: map['message'] as String? ?? 'An unknown error occurred',
        );
      default:
        // Unknown events are silently dropped — future-proofing
        return MeshErrorEvent(
          errorCode: 'UNKNOWN_EVENT',
          message: 'Unrecognised event type: $eventType',
        );
    }
  }
}

class PeerFoundEvent extends MeshEvent {
  final String peerId;
  final String peerName;
  final int rssi;
  final String connectionType;

  const PeerFoundEvent({
    required this.peerId,
    required this.peerName,
    required this.rssi,
    required this.connectionType,
  });
}

class PeerLostEvent extends MeshEvent {
  final String peerId;
  const PeerLostEvent({required this.peerId});
}

class PeerConnectedEvent extends MeshEvent {
  final String peerId;
  final String? groupOwnerIp;
  final bool isGroupOwner;

  const PeerConnectedEvent({
    required this.peerId,
    this.groupOwnerIp,
    this.isGroupOwner = false,
  });
}

class PeerDisconnectedEvent extends MeshEvent {
  final String peerId;
  const PeerDisconnectedEvent({required this.peerId});
}

class PacketReceivedEvent extends MeshEvent {
  final PacketModel packet;
  const PacketReceivedEvent({required this.packet});
}

/// Emitted both for global connection state and per-peer Wi-Fi Direct state changes.
/// [peerId] and [connectionState] are null for global broadcasts.
/// [connectionState] values: "connecting" | "connected" | "failed" | "disconnected"
class ConnectionStateChangedEvent extends MeshEvent {
  final bool isConnected;
  final int activePeersCount;
  final String? peerId;
  final String? connectionState;

  const ConnectionStateChangedEvent({
    required this.isConnected,
    required this.activePeersCount,
    this.peerId,
    this.connectionState,
  });
}

class MeshErrorEvent extends MeshEvent {
  final String errorCode;
  final String message;
  const MeshErrorEvent({required this.errorCode, required this.message});
}
